### PR TITLE
[SYCL] Add SYCL 2020 spec constants design doc

### PR DESCRIPTION
**Pull request by AlexeySachkov**

### CODE REVIEW

**Comment by AlexeySachkov** [comment]
This is a re-submit of #3331:
- moved the file to a new location
- adjusted a few other docs to point to the correct spec constants design
- made minor adjustments to FE section saying that `kernel_handler` argument is always generated

**Review by bader — DISMISSED** [review]

File: sycl/doc/design/SYCL2020-SpecializationConstants.md [hunk 1]
```diff
@@ -0,0 +1,1078 @@
...
+In order for runtime to access the right property, it need to compute the
+symbolic ID of a specialization constant based on user-provided inputs, such
+as non-type template argument passed to `set_specialization_constant` argument.
+DPC++ Headers section describes how symbolic IDs are generated and the same
+trick is used within `set_specialization_constant` method:
+```
+template<auto& SpecName>
+void set_specialization_constant(
+  typename std::remove_reference_t<decltype(SpecName)>::type value) {
+  const char *SymbolicID = detail::get_spec_constant_symbolic_ID();
```
Comment (gmlueck) [inline sycl/doc/design/SYCL2020-SpecializationConstants.md:966]:
```suggestion
  const char *SymbolicID = detail::get_spec_constant_symbolic_ID<SpecName>();
```

File: sycl/doc/design/SYCL2020-SpecializationConstants.md [hunk 2]
```diff
@@ -0,0 +1,1078 @@
...
+
+#include <CL/sycl/detail/spec_const_integration.hpp>
+```
+
+### DPC++ runtime
+
+For each device binary compiler generates a map
+\<Symbolic ID\> =\> \<list of spec constant descriptors\> ("ID map"). DPC++
+runtime imports that map when loading device binaries.
+It also maintains another map \<Spec const symbolic ID\> =\> \<its value\>
```
Comment (gmlueck) [inline sycl/doc/design/SYCL2020-SpecializationConstants.md:952]: Is the key the same thing in these two maps?  The ID map says the key is "\<Symbolic ID\>" while the value maps says the key is "\<Spec const symbolic ID\>".  If you intend for the key to be the same in both maps, use the description for both keys,  (For example, use "\<Symbolic ID\>" in both cases.)

File: sycl/doc/design/SYCL2020-SpecializationConstants.md [hunk 3]
```diff
@@ -0,0 +1,1078 @@
...
+pi_result piextProgramSetSpecializationConstant(pi_program prog,
+                                                pi_uint32 spec_id,
+                                                size_t spec_size,
+                                                const void *spec_value);
+```
+
+Plugin Interface function for descriptor of each property: `spec_id` and
+`spec_size` are taken from the descriptor, `spec_value` is calculated based on
+address of the specialization constant provided by user and `offset` field of
+the descriptor.
```
Comment (gmlueck) [inline sycl/doc/design/SYCL2020-SpecializationConstants.md:988]: > `spec_value` is calculated based on address of the specialization constant provided by user and `offset` field of the descriptor.

There is some missing step here.  How is `offset` used?  The value is in `specConstantValuesMap[SymbolicID]`.  Why not just pass the address of the entry in that map?

File: sycl/doc/design/SYCL2020-SpecializationConstants.md [hunk 4]
```diff
@@ -0,0 +1,1078 @@
...
+```
+
+the translator will generate `OpSpecConstant` SPIR-V instructions with proper
+`SpecId` decorations:
+
+```
+              OpDecorate %i32 SpecId 42 ; ID
+       %i32 = OpSpecConstant %int 0     ; Default value
+              OpDecorate %i32 SpecId 43                            ; ID of the 1st member
+              OpDecorate %float SpecId 44                          ; ID of the 2nd member
```
Comment (gmlueck) [inline sycl/doc/design/SYCL2020-SpecializationConstants.md:1062]:
```suggestion
              OpDecorate %A.i32 SpecId 43                            ; ID of the 1st member
              OpDecorate %A.float SpecId 44                          ; ID of the 2nd member
```

File: sycl/doc/design/SYCL2020-SpecializationConstants.md [hunk 5]
```diff
@@ -0,0 +1,1078 @@
...
+#endif __SYCL_DEVICE_ONLY__
+  }
+
+private:
+#ifdef __SYCL_DEVICE_ONLY__
+  template<auto &SpecName, typename T = std::remove_reference_t<decltype(SpecName)>::type>
+  // enable_if T is a scalar type
+  T get_on_device() {
+    auto ID = __builtin_sycl_unique_id(SpecName);
+    return __sycl_getScalar2020SpecConstantValue<T>(ID, &S, Ptr);
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply sycl/doc/design/SYCL2020-SpecializationConstants.md]: Yes, it should be `SpecName`, fixed in 197dd1981516596019382622275b1ff72874f4bb

File: sycl/doc/design/SYCL2020-SpecializationConstants.md [hunk 6]
```diff
@@ -0,0 +1,1078 @@
...
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+struct pi_device_binary_struct {
+...
+  // Array of property sets; e.g. specialization constants symbol-int ID map is
+  // propagated to runtime with this mechanism.
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+```
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply sycl/doc/design/SYCL2020-SpecializationConstants.md]: Removed in 0c8b59b7adc939362a4e02181d62090a1bf64b18

File: sycl/doc/design/SYCL2020-SpecializationConstants.md [hunk 7]
```diff
@@ -0,0 +1,1078 @@
...
+specialization constants is somewhere during or after linking device code from
+different translation units, so it happens in `sycl-post-link` tool.
+
+There is a `SpecConstantsPass` LLVM IR pass which:
+1. Assigns numeric IDs to specialization constants found in the linked module.
+2. Transforms IR to either:
+  a. The form expected by the SPIR-V translator (format of the
+     expected IR is covered in "Transformation of LLVM IR to SPIR-V friendly IR
+     form" section).
+  b. The form which is used for emulating specialization constants.
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply sycl/doc/design/SYCL2020-SpecializationConstants.md]: I've tried to fix that in 845353f97407072b281e25681007f4d695adaa2e - it now renders as a list, but not as a nested list. I will try another approach to make it a nested list as intended

File: sycl/doc/design/SYCL2020-SpecializationConstants.md [hunk 8]
```diff
@@ -0,0 +1,1078 @@
...
+...
+  // Array of property sets; e.g. specialization constants symbol-int ID map is
+  // propagated to runtime with this mechanism.
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+```
+
+So, within a single set we have a separate property for each specialization
+constant with name corresponding to its symbolic ID.
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply sycl/doc/design/SYCL2020-SpecializationConstants.md]: I've tried to address the comment in 7f64a62edc492777694860e91e4af9ae6cc26a6c - does it look better?

File: sycl/doc/design/SYCL2020-SpecializationConstants.md (hunk 2 above)
Comment (AlexeySachkov) [reply sycl/doc/design/SYCL2020-SpecializationConstants.md]: Yeah, those should refer to the same key, fixed in 7f64a62edc492777694860e91e4af9ae6cc26a6c

File: sycl/doc/design/SYCL2020-SpecializationConstants.md (hunk 3 above)
Comment (AlexeySachkov) [reply sycl/doc/design/SYCL2020-SpecializationConstants.md]: I've tried to explain that in more details in 737abfba9dd98f7bf27b439a242e4854d2d3b7c1. Does it look better?

File: sycl/doc/design/SYCL2020-SpecializationConstants.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/doc/design/SYCL2020-SpecializationConstants.md]: Yes, the nested list looks good now.  Which header are you talking about?

[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply sycl/doc/design/SYCL2020-SpecializationConstants.md]: There are couple of sub-sections following that list:
```
#### 2.1 Transformation of LLVM IR to SPIR-V friendly IR form
#### 2.2 Transformation of LLVM IR for emulating specialization constants
```

[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/doc/design/SYCL2020-SpecializationConstants.md]: Oh, I see.  I didn't even notice those numbers before or realize that they correspond to the list numbers.  I think it's fine the way it is.

**Review by gmlueck — APPROVED** [review]
LGTM!

**Comment by AlexeySachkov** [comment]
> buildbot/sycl-win-x64-pr Pending — Build started.

This check actually passed two days ago if you click [details](http://ci.llvm.intel.com:8010/#builders/18/builds/22454) link. @bader, I believe we can merge this

**Pull request merged**